activerecord/lib/active_record/delay_touching/state.rb

```diff
@@ -0,0 +1,71 @@
+module ActiveRecord
+  # = Active Record Delay Touching
+  module DelayTouching
+
+    # Tracking of the touch state. This class has no class-level data, so you can
+    # store per-thread instances in thread-local variables.
+    class State # :nodoc:
+      def initialize
+        @records = Hash.new { Set.new }
+        @already_updated_records = Hash.new { Set.new }
+      end
+
+      def updated(klass, attrs, records)
+        @already_updated_records[[klass, attrs]] += records
+      end
+
+      # Return the records grouped by class and columns that were touched:
+      #
+      #   {
+      #     [Owner, [:updated_at]]               => [owner1, owner2],
+      #     [Pet,   [:neutered_at, :updated_at]] => [pet1],
+      #     [Pet,   [:updated_at]]               => [pet2]
+      #   }
+      #
+      # As a side-effect, clears out the list of records.
+      def get_and_clear_records
+        records = @records
+        @records = Hash.new { Set.new }
+        records
+      end
+
+      def more_records?
+        @records.present?
+      end
+
+      def add_record(record, columns)
+        # Include the standard updated_at column and any additional specified columns
+        updated_at_attrs = record.send(:timestamp_attributes_for_update_in_model)
```

Comment:
Good catch. I can change `timestamp_attributes_for_update_in_model` to a public method with `:nodoc:`. That way it will be callable from outside the record, but not documented as a part of the public ActiveRecord API.